### PR TITLE
feat: weighted blending utilities for patch stitching

### PR DESCRIPTION
**Pull request by hkristen**
Second of three PRs splitting up #3214 (check out the archtitecture overview there for more context about how this will be used).

Depends on PR1 (GeoTIFFWriter #3666) -> set base branch to `feat/callbacks-geo-tiff-writer` for now. I will rebase onto main once #3666 merges.

Adds the blending and stitching layer used to reconstruct a full scene from overlapping inference patches. The core function is `weighted_merge()`, which reads patches from disk, applies per-patch cosine or linear blend masks (Hann window approach from https://doi.org/10.1371/journal.pone.0229839), and writes the merged result as a COG via `GeoTIFFWriter`. Patches are processed in chunks with a spatial grid index to keep memory usage flat regardless of scene size.

A few things worth noting: 
- patches at scene boundaries get their edge crop suppressed to avoid black borders (`_get_edge_deltas`), and the scene extent can either be inferred from patch transforms or pinned to the original dataset bounds. 
- The regression tests for both of those edge cases are in `TestBlackBorder` and `TestDatasetBoundsMode`.
- Also introduces PatchMetadata, the TypedDict that carries per-patch geo bounds and transform info between the callback and the merge step.

**ToDos for further versions / not included to reduce scope**
- [ ] BUG: When using delta=0, blending is fully disabled
- [ ] Widen output dtype to `uint16` when `num_classes > 255` to avoid label truncation

## AI Usage Disclosure

- [ ] 🟢 **No AI usage**: written by humans, for humans
- [x] 🟡 **AI-assisted**: AI helped with the coding, but I understand every line
- [ ] 🔴 **AI-generated**: AI did everything, review with caution

### CODE REVIEW

File: torchgeo/callbacks/blending.py
Comment (adriantre) [inline torchgeo/callbacks/blending.py:1]: Here you go! https://github.com/hkristen/torchgeo/pull/1

File: torchgeo/callbacks/blending.py [hunk 1]
```diff
@@ -0,0 +1,531 @@
...
+
+    top = 0 if abs(geo_bbox[3] - maxy) < tolerance else delta
+    bottom = 0 if abs(geo_bbox[1] - miny) < tolerance else delta
+    left = 0 if abs(geo_bbox[0] - minx) < tolerance else delta
+    right = 0 if abs(geo_bbox[2] - maxx) < tolerance else delta
+
+    return top, bottom, left, right
+
+
+def _reconstruct_scene_from_patches(
```
Comment (calebrob6) [inline torchgeo/callbacks/blending.py:106]: This won't work for "south-up" rasters like the AlphaEarth ones on source.coop. It always anchors the output at the maximum y-coordinate and assigns rows downward. This assumes a negative y-resolution.

File: torchgeo/callbacks/blending.py
Comment (hkristen) [reply torchgeo/callbacks/blending.py:1]: Thanks @adriantre - I included your suggested refactor in this branch :)

File: torchgeo/callbacks/blending.py (hunk 1 above)
Comment (hkristen) [reply torchgeo/callbacks/blending.py:106]: Good catch @calebrob6! I added support for "south-up" rasters -> it auto-detects the y_res sign and handles origin/row placement accordingly.
